src/SwedbankPay.Sdk/SwedbankPayHttpClient.cs

```diff
@@ -0,0 +1,227 @@
+        }
+
+
+        internal async Task<TResponse> HttpPost
+            <TPayLoad, TResponse>(string url, Func<ProblemsContainer, Exception> onError, TPayLoad payload)
+            where TResponse : new()
+        {
+            return await HttpRequest<TResponse>(HttpMethod.Post, url, onError, payload);
+        }
+
+
+        internal async Task<T> HttpRequest
+            <T>(string httpMethod, string url, Func<ProblemsContainer, Exception> onError, object payload = null)
+            where T : new()
+        {
+            return await HttpRequest<T>(new HttpMethod(httpMethod), url, onError, payload);
+        }
+
+
+        internal async Task<T> HttpRequest
+            <T>(HttpMethod httpMethod, string url, Func<ProblemsContainer, Exception> onError, object payload = null)
+            where T : new()
+        {
+            var msg = new HttpRequestMessage(httpMethod, url);
+
+            UpdateRequest(msg, payload);
+
+            HttpResponseMessage response;
+            try
+            {
+                response = await this.client.SendAsync(msg, HttpCompletionOption.ResponseHeadersRead).ConfigureAwait(false);
+            }
+            catch (HttpRequestException e)
+            {
+                throw new BadRequestException(e);
+            }
+            catch (TaskCanceledException te)
+            {
+                throw new ApiTimeOutException(te);
+            }
+
+            if (response.IsSuccessStatusCode)
+            {
+                var res = await response.Content.ReadAsStringAsync();
+                this.logger.LogInformation(res);
+                return JsonConvert.DeserializeObject<T>(res, JsonSerialization.JsonSerialization.Settings);
+            }
+
+            var responseMessage = await response.Content.ReadAsStringAsync();
+            this.logger.LogInformation(responseMessage);
+            ProblemsContainer problems;
+            if (!string.IsNullOrEmpty(responseMessage) && IsValidJson(responseMessage))
+                problems = JsonConvert.DeserializeObject<ProblemsContainer>(responseMessage);
+            else if (response.StatusCode == HttpStatusCode.NotFound)
+                problems = new ProblemsContainer("id", "Not found");
+            else
+                problems = new ProblemsContainer("Other", $"Response when calling SwedbankPay was: '{response.StatusCode}'");
+
+            var ex = onError(problems);
+            throw ex;
+        }
+
+
+        internal async Task<T> HttpRequest
+            <T>(HttpRequestMessage requestMessage, Func<ProblemsContainer, Exception> onError, object payload = null)
+            where T : new()
+        {
+            UpdateRequest(requestMessage, payload);
+
+            HttpResponseMessage response;
+            try
+            {
+                response = await this.client.SendAsync(requestMessage, HttpCompletionOption.ResponseHeadersRead).ConfigureAwait(false);
+            }
+            catch (HttpRequestException e)
```

Comment:
I feel these catch-clauses are obscuring the exception flow with unnecessary fluff. What's the difference between letting a TaskCanceledException bubble to the top instead of throwing it as an ApiTimeOutException? When the JSON parser is deserializing the content we don't catch and re-throw that, for example. I think the caller will have a better chance at understanding what went wrong if we allow the original exceptions to get thrown instead.